### PR TITLE
55677 Avoid duplicated query when retrieving empty collection from WP REST

### PR DESCRIPTION
**Pull request by gabriel-detassigny**
## The issue

**WP REST API** has this logic to re-run the query without the page parameter if we didn't get any posts back from the main one. This is to handle the case where we requested an extra page of a collection that is empty. In that case, we want the API to return an empty collection, but still return the total number of items in that collection. Therefore that extra query gives that total back.

However, it currently doesn't check if we were on the first page already. 
In that specific case, the second query would be the exact same as the first one and therefore wouldn't provide any additional value (the total of items would still be zero). This is why we sometimes ends up with a duplicated query.
By checking if we're already on the first page, we can avoid a duplicated query for empty collections.

Note that the trac ticket refers to the error happening on attachments endpoint, however any empty collection of posts will be affected (This is why some tests that were previously checking for two queries and had a "FIXME" comment, have been updated to one).

## How to test

- **Query Monitor** plugin should be enabled to detect any duplicate queries. 
- Open the gutenberg editor
- Open the JS console
- Type `await wp.apiFetch({ path: 'wp/v2/media?_envelope&media_type=text'}`
- The object returned should contain the number of duplicated queries. On trunc, there will be a duplicate which this patch will fix.

Trac ticket: [https://core.trac.wordpress.org/ticket/55677](https://core.trac.wordpress.org/ticket/55677)

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: tests/phpunit/tests/rest-api/rest-attachments-controller.php [hunk 1]
```diff
@@ -619,6 +619,33 @@ public function test_get_items_valid_modified_date() {
 		$this->assertSame( $id2, $data[0]['id'] );
 	}
 
+	/**
+	 * @ticket 55677
+	 */
+	public function test_get_items_avoid_duplicated_count_query_if_no_items() {
+		$request = new WP_REST_Request( 'GET', '/wp/v2/media' );
+		$request->set_param( 'media_type', 'video' );
+		$initial_num_queries = get_num_queries();
```
Comment (TimothyBJacobs) [inline tests/phpunit/tests/rest-api/rest-attachments-controller.php]: Tests that verify `num_queries` have the potential to be brittle. Could we update this to use similar logic as the tests you updated? That is checking for the actual `WP_Query`s that are run. For instance by checking the number of time `post_clauses` is executed.

Comment (gabriel-detassigny) [reply tests/phpunit/tests/rest-api/rest-attachments-controller.php]: Thanks @TimothyBJacobs for the feedback. I've now changed these tests to use `post_clauses` as you suggested

Comment (TimothyBJacobs) [reply tests/phpunit/tests/rest-api/rest-attachments-controller.php]: Looks great, thanks!

**Comment by TimothyBJacobs** [comment]
Can we also interrogate the `X-WP-TotalPages` and `X-WP-Total` headers in our tests here to have assertions that they still match our expected results.

**Comment by gabriel-detassigny** [comment]
Thanks @TimothyBJacobs . I've made some changes now:
- first test for an empty collection now also asserts these headers
- second test checks the appropriate _invalid page number_ error is returned (headers cannot be verify because they don't get set in this specific case, [the error is returned right before the headers would be set](https://github.com/WordPress/wordpress-develop/blob/2f8a627bd16f615d7b5775f56b7347af6a679b39/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L402-L413))

**Comment by TimothyBJacobs** [comment]
Thanks @gabriel-detassigny, this looks good to me!

**Review by spacedmonkey — APPROVED** [review]

**Comment by spacedmonkey** [comment]
Committed https://github.com/WordPress/wordpress-develop/commit/219927c11b001fad95e1f2ee6b17d132b8d00af2.

**Pull request closed without merging**